packages/truffle-decoder/lib/contract.ts

```diff
@@ -43,156 +43,92 @@ export default class TruffleContractDecoder extends AsyncEventEmitter {
       this.contextHash = hash;
-      this.contexts[hash] = context;
+      this.context = this.contexts[hash];
     }
-    if(this.contract.bytecode) { //now the constructor version
-      const constructorContext = Utils.makeContext(this.contract, this.contractNode, true);
-      debug("adding context: %O", constructorContext);
+    if(this.contract.bytecode && this.contract.bytecode !== "0x") { //now the constructor version
       const hash = CodecUtils.Conversion.toHexString(
         CodecUtils.EVM.keccak256({type: "string",
-          value: constructorContext.binary
+          value: this.contract.bytecode
         })
       );
-      debug("with hash: %s", hash);
-      this.constructorContext = constructorContext;
       this.constructorContextHash = hash;
-      this.contexts[hash] = constructorContext;
+      this.constructorContext = this.contexts[hash];
     }
 
-    for(let relevantContract of this.relevantContracts) {
-      let node: AstDefinition = Utils.getContractNode(relevantContract);
-      if(node !== undefined) {
-        this.contracts[node.id] = relevantContract;
-        this.contractNodes[node.id] = node;
-        if(relevantContract.deployedBytecode) {
-          const context = Utils.makeContext(relevantContract, node);
-          debug("adding context: %O", context);
-          const hash = CodecUtils.Conversion.toHexString(
-            CodecUtils.EVM.keccak256({type: "string",
-              value: context.binary
-            })
-          );
-          this.contexts[hash] = context;
-        }
-      }
-    }
-
-    debug("contexts: %o", this.contexts);
-    this.contexts = <CodecUtils.Contexts.DecoderContexts>CodecUtils.Contexts.normalizeContexts(this.contexts);
-    this.context = this.contexts[this.contextHash];
-    this.constructorContext = this.contexts[this.constructorContextHash];
-    this.contextsById = Object.assign({}, ...Object.values(this.contexts).filter(
-      ({isConstructor}) => !isConstructor
-    ).map(context =>
-      ({[context.contractId]: context})
-    ));
-  }
-
-  public async init(): Promise<void> {
-    this.contractNetwork = (await this.web3.eth.net.getId()).toString();
-    if(this.contractAddress === undefined) {
-      this.contractAddress = this.contract.networks[this.contractNetwork].address;
-    }
-
-    debug("init called");
-    [this.referenceDeclarations, this.userDefinedTypes] = this.getUserDefinedTypes();
-
-    let libraryAllocationInfo: Codec.ContractAllocationInfo[] =
-    Object.entries(this.contracts).filter(
-      ([id, _]) => this.contractNodes[parseInt(id)].contractKind === "library"
-    ).
-    map(
-      ([id, { abi }]) => ({
-        abi: AbiUtils.schemaAbiToAbi(abi),
-        id: parseInt(id)
-      })
-    );
+    this.referenceDeclarations = this.wireDecoder.getReferenceDeclarations();
+    this.userDefinedTypes = this.wireDecoder.getUserDefinedTypes();
 
     this.allocations = {};
+    this.allocations.abi = this.wireDecoder.getAbiAllocations();
```

Comment:
Same here - is this data duplication necessary?

Reply:
This one I'd prefer to keep, because I'd like to be able to just set `this.allocations` once and then be done with it.